Jenkinsfile

```diff
@@ -88,116 +270,250 @@ def step_compile_results()
         if ("${params.HIL_RIOT_VERSION}" == 'pull' && "${params.HIL_RIOT_PULL}" != '0') {
             // checkout specified PR number
             def prnum = params.HIL_RIOT_PULL.toInteger()
-            sh """
+            sh script: """
                 cd RIOT
                 git fetch origin +refs/pull/${prnum}/merge
                 git checkout FETCH_HEAD
-            """
+            """, label: "checkout RIOT PR ${prnum}"
         }
     }
 }
 
+/* Gets boards to test based on parameters or dynamically all available
+ * boards connected to jenkins nodes.
+ *
+ * Sets nodeBoards
+ */
+def stepGetBoards() {
+    if (params.BOARDS == 'all') {
+        nodeBoards = getBoardsFromNodesEnv()
+    }
+    else {
+        nodeBoards = params.BOARDS.tokenize(',')
+        /* TODO: Validate if the boards are connected */
+    }
+    sh script: "echo collected boards: ${nodeBoards.join(",")}",
+            label: "print boards"
+}
+
+/* Gets unique boards connected to jenkins nodes. */
+def getBoardsFromNodesEnv() {
+    script {
+        boards = []
+        for (int i=0; i < nodes.size(); ++i) {
+            node (nodes[i]) {
+                boards.push(env.BOARD)
+            }
+        }
+        boards.unique()
+        return boards
+    }
+}
+
+/* Gets test based on parameters or dynamically all available
+ * tests in RobotFW-Tests.
+ *
+ * Sets nodeTests
+ */
+def stepGetTests() {
+    if (params.TESTS == 'all') {
+        nodeTests = getTestsFromDir()
+    }
+    else {
+        nodeTests = params.TESTS.tokenize(',')
+    }
+
+    sh script: "echo collected tests: ${nodeTests.join(",")}",
+            label: "print tests"
+}
+
+/* Gets tests in tests directory. */
+def getTestsFromDir() {
+    script {
+        tests = sh returnStdout: true,
+                script: """
+                    for dir in \$(find tests -maxdepth 1 -mindepth 1 -type d); do
+                        [ -d \$dir/tests ] && { echo \$dir ; } || true
+                    done
+                """
+        tests = tests.tokenize()
+        return tests
+    }
+}
+
+/* Iterates through each board in nodeBoards and test in nodeTests and builds. */
+def stepBuildJobs() {
+    script {
+        for (int t_idx=0; t_idx < nodeTests.size(); t_idx++) {
+            for (int b_idx=0; b_idx < nodeBoards.size(); b_idx++) {
+                buildJob(nodeBoards[b_idx], nodeTests[t_idx])
+            }
+        }
+    }
+}
+
+/* Iterates through each board in nodeBoards and test in nodeTests and builds
+ * then stashes successful build binaries.
+ *
+ * For example, if a "board=samr21-xpro" and the test is "tests/periph_gpio",
+ * the binaries (.hex, .bin, .elf) will be stashed in
+ * "samr21_xpro_tests_periph_gpio".
+ *
+ * @param board The board to build
+ * @param test  The test to build
+ */
+def buildJob(board, test) {
+    exit_code = sh script: "${params.EXTRA_BUILD_ARGS} BUILD_IN_DOCKER=1 BOARD=${board} make -C ${test} clean all -j",
```

Comment:
Note that there is the issue https://github.com/RIOT-OS/RIOT/issues/14504 when passing extra build args to docker. There is also a workaround linked there but I'm not sure how that applies here.

Reply:
I made it more generic, We also don't have the proper build server setup so there will probably be some tuning after the fact.